doc/src/documentation_style_guide.rst

```diff
@@ -0,0 +1,1019 @@
+
+Sample Docstring
+================
+
+Here is an example of a correctly formatted docstring::
+
+  class gamma(Function):
+      r"""
+      The gamma function
+
+      .. math::
+          \Gamma(x) := \int^{\infty}_{0} t^{x-1} e^{-t} \mathrm{d}t.
+
+      Explanation
+      ===========
+
+      The ``gamma`` function implements the function which passes through the
+      values of the factorial function (i.e., `\Gamma(n) = (n - 1)!`), when n is
+      an integer. More generally, `\Gamma(z)` is defined in the whole complex
+      plane except at the negative integers where there are simple poles.
+
+      Examples
+      ========
+
+      >>> from sympy import S, I, pi, oo, gamma
+      >>> from sympy.abc import x
+
+      Several special values are known:
+
+      >>> gamma(1)
+      1
+      >>> gamma(4)
+      6
+      >>> gamma(S(3)/2)
+      sqrt(pi)/2
+
+      The Gamma function obeys the mirror symmetry:
+
+      >>> from sympy import conjugate
+      >>> conjugate(gamma(x))
+      gamma(conjugate(x))
+
+      Differentiation with respect to x is supported:
+
+      >>> from sympy import diff
+      >>> diff(gamma(x), x)
+      gamma(x)*polygamma(0, x)
+
+      Series expansion is also supported:
+
+      >>> from sympy import series
+      >>> series(gamma(x), x, 0, 3)
+      1/x - EulerGamma + x*(EulerGamma**2/2 + pi**2/12) + x**2*(-EulerGamma*pi**2/12 +
+      polygamma(2, 1)/6 - EulerGamma**3/6) + O(x**3)
+
+      We can numerically evaluate the gamma function to arbitrary precision
+      on the whole complex plane:
+
+      >>> gamma(pi).evalf(40)
+      2.288037795340032417959588909060233922890
+      >>> gamma(1+I).evalf(20)
+      0.49801566811835604271 - 0.15494982830181068512*I
+
+      See Also
+      ========
+
+      lowergamma: Lower incomplete gamma function.
+      uppergamma: Upper incomplete gamma function.
+      polygamma: Polygamma function.
+      loggamma: Log Gamma function.
+      digamma: Digamma function.
+      trigamma: Trigamma function.
+      sympy.functions.special.beta_functions.beta: Euler Beta function.
+
+      References
+      ==========
+
+      .. [1] https://en.wikipedia.org/wiki/Gamma_function
+      .. [2] http://dlmf.nist.gov/5
+      .. [3] http://mathworld.wolfram.com/GammaFunction.html
+      .. [4] http://functions.wolfram.com/GammaBetaErf/Gamma/
+
+      """
+
+Docstrings for Classes that are Mathematical Functions
+======================================================
+
+SymPy is unusual in that it also has classes that are mathematical functions.
+The docstrings for classes that are mathematical functions should include
+details specific to this kind of class, as noted below:
+
+* The Explanation section should include a mathematical definition of the
+  function. This should use LaTeX math. Use `$$ for inline math
+  <https://github.com/sympy/sympy/wiki/SymPy-Documentation-Style-Guide#formatting-
+  preferences>`_ and .. math:: for display math, which should be used for the
```

Comment:
I think suggesting double dollars for math is bad for two reasons:

* double dollars are very much discouraged for use in LaTeX proper, using them here would set a bad example
* double dollar math can be used in the middle of a paragraph (right?), which allows very messy and unstructured docstrings. Using only the `.. math::` directive would force display math to a separate line, which would be much cleaner and easier to read in the source code.

Reply:
Double dollars for display math are pretty common in markup languages though, like the Jupyter notebook and other places that use Markdown like Math SE. 

If you have any comments on ways we can improve the extension, please comment over there.

Reply:
I don't know if they are common, but I know that there are people who use double dollars.

But that's bad, and we should not help spread this bad habit.

> If you have any comments on ways we can improve the extension, please comment over there.

The extension is fine, people should just not use double dollars.
And the SymPy style guide should not encourage them to do so.

Reply:
> The extension is fine, people should just not use double dollars.

This seems to contradict itself, once I merge the linked PR. 

> double dollar math can be used in the middle of a paragraph (right?)

The way I have it now yes, but this can easily be modified. Again, please comment these kinds of things on the PR, or open issues on the extension.

Reply:
> > The extension is fine, people should just not use double dollars.
>
> This seems to contradict itself, once I merge the linked PR.

It doesn't. Just because something is possible doesn't mean that it's a good idea to do it.

> Again, please comment these kinds of things on the PR, or open issues on the extension.

Again, I have no problem with the extension.
People should simply not use double dollars. Whether the extension supports them or not.

Reply:
I'm writing the extension specifically for SymPy, but making it available so that others can use it as well. So if we want to limit something, we should add support directly in the extension. It's quite easy to restrict which dollars are parsed by updating the regular expression. And it's also quite easy to add configuration options to the extension.

Reply:
I agree with @mgeier that it would be perfectly ok to support double-dollar in the extension but **these guidelines** can nonetheless ignore it (or actively discourage as you wish).